x10.1
.TH X10 1 local
.SH NAME
.B x10\^
- control program for the BSR X10 Powerhouse
.SH SYNOPSIS
.B x10 data \fIunit-name\fP \fIbinary-state\fP \fIid\fP
.br
.B x10 date
.br
.B x10 delete \fI[event|data]\fP \fIslot-numbers\fP
.br
.B x10 diagnostic
.br
.B x10 dump \fI[event|data|all]\fP
.br
.B x10 fdump \fI[event|data]\fP
.br
.B x10 finfo \fI[event|data]\fP
.br
.B x10 fload \fI[event|data]\fP
.br
.B x10 info
.br
.B x10 turn \fIunit-specifier\fP \fIstate\fP
.br
.B x10 monitor
.br
.B x10 reset [\fIhousecode\fP]
.br
.B x10 schedule \fIunit-specifier\fP \fIday-specifier\fP \fItime\fP \fIstate\fP \fI[slot-number]\fP
.br
.B x10 setclock
.SH DESCRIPTION
.I x10
is a little program for controlling a BSR X10 "Powerhouse" home
control device.  The BSR X10 can remotely control lights and appliances
in your house by signalling over the house wiring.
.PP
The BSR X10 Powerhouse connects to a computer with an RS232 interface.
It can store up to 128 events;  each event can turn on, turn off, or dim
up to sixteen slave units.  The X10 box has a battery backed up clock
which the computer can read.
.PP
Slave units are identified by a one-letter housecode ranging from A to P (for 16
different codes) and a number from 1 to 16, for a total of 256 possible
unit codes.
.PP
You could just put a bunch of "x10 turn" commands in your crontab, but
this doesn't work if your system is down for backups, or has crashed,
or if someone's tripped over the RS232 cable and unplugged it, and it
clutters up your crontab something awful.  For most uses, it's much
easier to just load the events into the X10.  You can dump the schedule
into a file (with the \fBx10 fdump\fP command),
and you can have different files for different times of
the year, for before and after Daylight Savings Time, etc.
.PP
X10 unit descriptions are kept in a file called \fBid\fR.  Each
line contains a number (1-255) and a unit description (0-39 characters),
separated by a tab.
The number represents a code that can be stored in the X10 along
with a unit id using the \fBx10 data\fR command.  If the \fBid\fR
file exists, the \fBx10 dump data\fR or \fBx10 dump all\fR commands
will print descriptions for each unit code contained in the file.
.SH COMMANDS
.IP \fBdata\fP \fIunit-name\fP \fIbinary-state\fP \fIid\fP
Loads data into "graphics" storage in X10 box.  Currently used to
record correspondence between id file identifier and unit code,
and current state of unit.
This stuff is incomplete at present.
.PP
.IP \fBdate\fP
Gets current date and time from X10 and displays it in a form suitable
for feeding to \fIdate(1)\fP as input.  Thus, you can set your system clock
by saying "date `x10 date`".
.PP
.IP \fBdelete\fP \fI[event|data]\fP \fIslot-numbers\fP
Deletes the request of the given type identified by the given
\fIslot-numbers\fP from X10 memory.  \fIslot-numbers\fP can be either
individual numbers or hyphen-separated ranges.
.PP
.IP \fBdiagnostic\fP
Runs the X10 diagnostic and reports the result.  \fBCaution:\fP running
the diagnostic scrambles X10 memory, losing all downloaded events and
data.
.PP
.IP \fBdump\fP \fI[event|data|all]\fP
Writes a readable description of the requested event types
on standard output.
.PP
.IP \fBfdump\fP \fI[event|data]\fP
Writes an unreadable description (binary data) of all events of type
\fIevent-type\fP on standard output.  If you redirect this stuff to
a file, you can subsequently feed it to a \fBx10 fload\fP command.
.PP
.IP \fBfinfo\fP \fI[event|data]\fP
Like \fBdump\fP, but reads binary data in \fBfdump\fP format from
standard input, rather than reading data from the X10.
.PP
.IP \fBfload\fP \fI[event|data]\fP
Loads the X10 memory with requests of the given type from
data on standard input in \fBx10 fdump\fP format.
.PP
.IP \fBinfo\fP
Displays current setting of X10's clock and housecode.
.PP
.IP \fBturn\fP \fIunit-specifier\fP \fIstate\fP
Sets the specified unit to the specifed state.
.PP
.SH ARGUMENTS
Most X10 commands require arguments.  They are as follows:
.TP 5
\fIhousecode\fP
housecode is a letter from a-p
.TP 5
\fIid\fP
The identification number of the unit (1-255).  It is used to associate
a description string from the ID file to a specific unit.
.TP 5
\fIslot-number\fP
An X10 event index (0-127) for event requests,  or id-slot (0-255) for
data requests.
.TP 5
\fIbinary-state\fP
id-state is either \fBon\fR or \fBoff\fR
.TP 5
\fIday-specifier\fP
mode-day is in the form: (mode) ([day])
.br
Modes:
.RS 5
 \fBn\fRormal <days>
.br
	\fBse\fRcurity <days>
.br
	\fBtod\fRay
.br
	\fBtom\fRorrow
.br
The "security" feature is implemented in the CP-290 itself, and applies
some variability to the specified time.
.RE
Days:
.RS 5
\fBm\fRonday, \fBtu\fResday, \fBwed\fRnesday,
\fBth\fRursday, \fBf\fRriday,
.br
\fBsa\fRturday, \fBsu\fRnday, \fBweekd\fRay,
\fBweeke\fRnd, \fBe\fRveryday
.RE
.TP 5
\fIunit-name\fP
A name for a single unit in the form: (housecode)(unit-number)
.RS 5
housecode is a letter from a-p
unit-number is a number from 1-16
.RE
Or, if the x10config file is present, and contains aliases, the unit-name
may be one of those aliases.
.TP 5
\fIunit-specifier\fP
A specifier for one or more units, in the form
(housecode)(unit-number[,unit-number...])
.RS 5
housecode is a letter from a-p
.br
unit-number is a number from 1-16 or a '*' to indicate all units
.RE 3
.TP 5
\fIslot-number\fP
slot-numbers have a range of 0-127 for events, and 0-255 for data.
.br
request-type is \fBevents\fR or \fBdata\fR or \fBall\fR for dump command
.TP 5
\fIstate\fP
state is in the form: (function) ([value])
.RS 5
.B on
.br
.B off
.br
.B dim
(value)
.RS 5
value is a number from 1-16
.RE 3
.TP 5
\fItime\fP
Time is in 24 hour format of the form HH:MM, or "\fBsunset\fP" or "\fBsunrise\fP"
with an optional trailing offset, e.g. sunset+1:00 or sunrise-0:45. 
.SH EXAMPLES
.TP 5
x10 turn a5 on
Turns slave unit A5 on.
.TP 5
x10 turn b7 dim 8
Sets unit B7 to dim level 8 (levels run from 0 to 15, bright to dim)
.TP 5
x10 info
Displays X10 clock time and base housecode
.TP 5
x10 setclock
Sets X10's clock to current time of day (from UNIX)
.TP 5
x10 schedule a3,5 security everyday 17:30 off
Enters an event to turn units A3 and A5 off every day at approximately 17:30.
.TP 5
x10 schedule frontporch normal everyday sunset+0:30 off
Enters an event to turn the aliased lights off every day half an hour
after sunset.
.TP 5
x10 schedule a2 normal wed 5:00 on 14
Enters an event to turn unit A2 on every Wednesday at 5:00, and store
the event in slot 14.  If the '14' is left off, the event will be stored
in the first free slot.  It is much faster to tell it the slot number if
programming many events at once.
.TP 5
x10 schedule b8 today 9:00 dim 9
Enters an event to dim unit B8 to level 9 today at 9:00
.TP 5
x10 dump all
Displays all events and data in X10's memory
.TP 5
x10 delete event 12
Deletes event 12 from X10's memory
.TP 5
x10 reset
Clears X10 memory, deletes all events and sets housecode A
.br
(does not clobber clock)
.TP 5
x10 reset c
Clears X10 memory, deletes all events and sets housecode C
.br
(does not clobber clock)
.TP 5
x10 diagnostic
The X10 performs its internal diagnostic routine
.br
(clobbers memory and clock, but not housecode)
.TP 5
x10 fdump data > datafile
Dumps X10 id data into file
.TP 5
x10 fload events <eventfile
Loads X10 event data from file
.TP 5
x10 finfo events < eventfile
Displays events in file produced by x10 fdump
.TP 5
x10 date
Displays date in date(1) input format.
.SH FILES
id - X10 unit description file
.SH AUTHORS
Originally written by Larry Campbell (maynard!campbell).  System V
port, ID file, improved display formats, and other cleanup by John
Chmielewski (rogue!jlc).  Module aliasing, sunset and sunrise
additions by Paul Fox (pgf@foxharp.boston.ma.us)
.SH SEE ALSO
X10 POWERHOUSE computer interface model no. CP290 programming guide
.br
date(1)


x10.c

/*
 * Copyright 1986 by Larry Campbell, 73 Concord Street, Maynard MA 01754 USA
 * (maynard!campbell).  You may freely copy, use, and distribute this software
 * subject to the following restrictions:
 *
 *  1)	You may not charge money for it.
 *  2)	You may not remove or alter this copyright notice.
 *  3)	You may not claim you wrote it.
 *  4)	If you make improvements (or other changes), you are requested
 *	to send them to me, so there's a focal point for distributing
 *	improved versions.
 *
 * John Chmielewski (tesla!jlc until 9/1/86, then rogue!jlc) assisted
 * by doing the System V port and adding some nice features.  Thanks!
 */

#include <stdio.h>
#include <stdlib.h>
#include <ctype.h>
#ifdef SYSV
#include <string.h>
#else
#include <strings.h>
#endif
#include <time.h>
#include "x10.h"

extern long time();
extern struct tm *localtime();
extern struct nstruct modnames[];
extern int tty;

int x10_housecode;

void sigtimer();
char hc2char();

char
 syncmsg[SYNCN], flag;

char latitude[20];
char longitude[20];

struct hstruct			/* table to map housecodes into letters */
 housetab[] =
{
    {HC_A, 'a'},
    {HC_B, 'b'},
    {HC_C, 'c'},
    {HC_D, 'd'},
    {HC_E, 'e'},
    {HC_F, 'f'},
    {HC_G, 'g'},
    {HC_H, 'h'},
    {HC_I, 'i'},
    {HC_J, 'j'},
    {HC_K, 'k'},
    {HC_L, 'l'},
    {HC_M, 'm'},
    {HC_N, 'n'},
    {HC_O, 'o'},
    {HC_P, 'p'}
};

char *wdays[] =
{"Sunday", "Monday", "Tuesday", "Wednesday",
 "Thursday", "Friday", "Saturday", "<day not set>"};

unsigned char	/* table to map unit numbers into unit bit mask */
 maphibyt[] =
{
    0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
    0x80, 0x40, 0x20, 0x10, 0x08, 0x04, 0x02, 0x01
}, maplobyt[] =
{
    0x80, 0x40, 0x20, 0x10, 0x08, 0x04, 0x02, 0x01,
    0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00
};

int
 timeout = TIMEOUT, Iloaded, Iminutes, Ihours, Idays;

unsigned char
 Ihcode;

extern int
 c_data(), c_date(), c_delete(), c_diagnostic(), c_dump(), c_fdump(),
 c_finfo(), c_fload(), c_info(), c_monitor(), c_reset(), c_schedule(),
 c_setclock(), c_turn();

struct cmdentry {
    char *cmd_name;
    int (*cmd_routine) ();
} cmdtab[] = {
    "data", c_data,
    "date", c_date,
    "delete", c_delete,
    "diagnostic", c_diagnostic,
    "dump", c_dump,
    "fdump", c_fdump,
    "finfo", c_finfo,
    "fload", c_fload,
    "info", c_info,
    "monitor", c_monitor,
    "reset", c_reset,
    "schedule", c_schedule,
    "setclock", c_setclock,
    "turn", c_turn,
    "", NULL
};

main(argc, argv)
char *argv[];
{
    register i;
    int (*rtn) ();
    struct cmdentry *c;

    if (argc < 2)
	usage(E_NOCMD);

    rtn = NULL;
    for (c = cmdtab; c->cmd_routine != NULL; c++) {
	if (strcmp(argv[1], c->cmd_name) == 0) {
	    rtn = c->cmd_routine;
	    break;
	}
    }
    if (rtn == NULL)
	usage(E_INVCN);

    read_config();

    setup_tty();

#ifdef MINIEXCH
    mxconnect(MINIXPORT);
#endif

    for (i = 0; i < SYNCN; i++)
	syncmsg[i] = i < 11 ? 0xEF : 0xFF;
    init();
    (*rtn) (argc, argv);
    restore_tty();
    close_tty();
    return 0;
}

/*
 * Convert X10-style day of week (bit map, bit 0=monday, 6=sunday)
 * to UNIX localtime(3) style day of week (integer, 0=sunday)
 */

dowX2U(b)
register char b;
{
    register n;

    for (n = 1; (!(b & 1)) && n < 8; n++, b = b >> 1)
    	;
    if (n == 7)
	n = 0;
    if (n == 8)
	n = 7;
    return (n);
}

dowU2X(d)
register d;
{
    if (d == 0)
	d = 7;
    return (1 << (d - 1));
}

init()
{
    int n;
    unsigned char buf[6];

    sendsync();
    usleep(100000);
    sendsync();
    usleep(100000);
    (void) write(tty, "\004", 1);	/* GETINFO command */
    getsync();
    n = xread(tty, buf, 6, timeout);

    if (n != 6) {
	fprintf(stderr,"invalid Clock and Base Housecode message length\n");
	fprintf(stderr,"n = %d, buf %02x %02x %02x %02x %02x %02x\n",
	    n, buf[0],buf[1],buf[2],buf[3],buf[4],buf[5]);
	quit();
    }
    if (CHKSUM(buf) != buf[5])
	error("checksum error");

    Iloaded = buf[0] & 1;
    Iminutes = buf[1];
    Ihours = buf[2];
    Idays = buf[3];
    Ihcode = buf[4];
}

chkack()
{
    unsigned char buf[7];
    int n;

    n = xread(tty, buf, 7, timeout);
    if (n != 7) {
	int i;

	(void) fprintf(stderr, "chkack dump (%d bytes):\n", n);
	for (i = 0; i < n; i++)
	    (void) fprintf(stderr, "buf[%d] = 0x%x\n", i, buf[i]);
	error("timeout while awaiting ACK message");
    }
}

/*
 * Check command report ("Command Upload", the manual calls it).
 * If argument supplied is non-zero, print the report in human-readable
 * form.
 */

chkrpt(printflag)
{
    static char *statetab[] =
    {"S0", "S1", "ON", "OFF", "DIM", "DIM", "S6", "S7"};
    int n;
    unsigned char buf[6];
    long dtime;
    struct tm *tp;

    getsync();
    n = xread(tty, buf, 6, timeout);

    if (n != 6)
	error("chkrpt: invalid event report length");
    if (CHKSUM(buf) != buf[5]) {
	(void) fprintf(stderr,
	"Checksum computed: 0x%x, received: 0x%x\n", CHKSUM(buf), buf[5]);
	error("chkrpt: checksum error");
    }
    if (!printflag)
	return;

    dtime = time((long *) 0);
    tp = localtime(&dtime);

    (void) printf("%2d:%02d:%02d: unit ",
	    tp->tm_hour, tp->tm_min, tp->tm_sec);
    punits(hc2char(buf[1] & 0xF0), (buf[3] << 8) | buf[2]);
    (void) printf(", state %s\n", statetab[buf[1] & 0x07]);
}

getsync()
{
    unsigned char buf[RCVSYNC];
    int n;

    if ((n = xread(tty, buf, RCVSYNC, timeout)) < RCVSYNC) {
	fprintf(stderr,"got %d chars\n",n);
#if 0
	fprintf(stderr,"syn %02x %02x %02x %02x %02x %02x\n",
	    buf[0],buf[1],buf[2],buf[3],buf[4],buf[5]);
#endif
	error("Failed to get sync characters");
    }
#if 0
    printf("syn %02x %02x %02x %02x %02x %02x\n",
	buf[0],buf[1],buf[2],buf[3],buf[4],buf[5]);
#endif
}

sendsync()
{
    (void) write(tty, syncmsg, SYNCN);
}

chksum(buf, size)
unsigned char *buf;
{
    register i, sum;

    for (i = 1, sum = 0; i < (size - 1); i++)
	sum += buf[i];
    return (sum & 0xFF);
}

char hc2char(code)
unsigned code;
{
    register i;

    for (i = 0; i < 16; i++)
	if (housetab[i].h_code == code)
	    return i + 'a';
    return ('?');
}

char2hc(hletter)
int hletter;
{
    if (isupper(hletter))
	hletter = tolower(hletter);
    if ('a' <= hletter && hletter <= 'p')
        return housetab[hletter - 'a'].h_code;
    else
	error("invalid house code");
}

/*
 * Parse string of comma-separated unit numbers and return bitmap
 * (big-endian) of units specified.  '*' means "all units".
 */

getunits(p)
register char *p;
{

#define DIGBUFN 80

    unsigned lobits, hibits, n, unit, lastunit = 0;
    char digbuf[DIGBUFN];

    lobits = 0;
    hibits = 0;
    while (*p) {
	if (*p == '*') {
	    lobits = 0xFF;
	    hibits = 0xFF;
	    break;
	}
	for (n = 0; n < DIGBUFN && isdigit(*p); n++, p++)
	    digbuf[n] = *p;
	digbuf[n] = '\0';
	if ((unit = atoi(digbuf)) < 1 || unit > 16)
	    error("bad unit code, must be between 1 and 16");
	if (lastunit == 0)
		lastunit = unit;
	else if (lastunit > unit) {
		n = unit; unit = lastunit; lastunit = n;
	}

	for (n = lastunit; n <= unit; n++) {
		lobits |= maplobyt[n - 1];
		hibits |= maphibyt[n - 1];
	}
	if (*p)
	    if (*p == '-') {
		lastunit = unit;
		p++;
	    } else if (*p == ',') {
		lastunit = 0;
		p++;
	    } else {
		error("bad unit separator, use comma or dash please");
	    }
    }
    return ((lobits << 8) | hibits);
}

dimstate(p, level)
register char *p, *level;
{
    unsigned levelnum;

#if BEFORE
    if (strcmp(p, "on") == 0)
	return (2);
    if (strcmp(p, "off") == 0)
	return (3);
    if (strcmp(p, "dim") != 0)
	error("bad state keyword");
    if (sscanf(level, "%d", &levelnum) == 0)
	error("dim value must be numeric");
    if (levelnum > 15)
	error("dim value out of range, must be between 0 and 15");
    timeout = DTIMEOUT;
    return ((levelnum << 4) | 5);
#else
    int cmd;

    if (strcmp(p, "on") == 0)
	return (2);
    if (strcmp(p, "off") == 0)
	return (3);
    if (strcmp(p, "dim_to_off") == 0)
	return (4);

    if (strcmp(p, "dim") == 0 || strcmp(p, "on_and_dim") == 0)
	cmd = 5;
    else if (strcmp(p, "brighten") == 0)
	cmd = 6;
    else
	error("bad state keyword: on/off/dim(on on_and_dim)/dim_to_off/brighten");
    if (sscanf(level, "%d", &levelnum) == 0)
	error("level value must be numeric");
    if (levelnum > 15)
	error("dim value out of range, must be between 0 and 15");
    timeout = DTIMEOUT;
    return ((levelnum << 4) | cmd);
#endif
}

struct nstruct dtab[] =
{
    "monday", 0x01,
    "tuesday", 0x02,
    "wednesday", 0x04,
    "thursday", 0x08,
    "friday", 0x10,
    "saturday", 0x20,
    "sunday", 0x40,
    "everyday", 0x7f,
    "weekdays", 0x1f,
    "weekends", 0x60,
    "", 0x00
};

day2bits(p)
char *p;
{
    char c, buf[6];
    int n, mask, length;

    if (strcasecmp(p,"Today") == 0)
    	return (Idays);
    length = strlen(p);
    mask = 0;
    for (n = 0; dtab[n].n_name[0] != 0; n++) {
	if (strncasecmp(dtab[n].n_name, p, length) == 0) {
	    if (mask != 0)
		error("ambiguous day abbreviation");
	    mask = dtab[n].n_code;
	}
    }
    if (mask == 0)
	error("bad day keyword");
    return (mask);
}

mode2code(p)
char *p;
{
    int len = strlen(p);
    int n;
    for (n = 0; modnames[n].n_name[0]; n++) {
	if (strncasecmp(modnames[n].n_name, p, len) == 0) {
	    flag = n;	/* position of function name in table */
	    return modnames[n].n_code;
	}
    }
    error("bad mode keyword");
}

#define MODULES 100
#define NAME_LEN 20
#define UNIT_LEN 50

struct x10_mod {
	char name[NAME_LEN+1];
	char alias[NAME_LEN+1];
	char hc;
	char un[UNIT_LEN];
} x10_modules[MODULES];


read_config()
{

	char line[256];
	char alias[256];
	char housecode[256];
	char unit[256];
	char *configfile;
	char *comment;
	static FILE *f = NULL;
	int i = 0;
	int n;
	extern char *getenv();

	if (f) return;

	configfile = getenv("X10CONFIG");
	if (!configfile) {
		(void) strcat(strcpy(line, XDIR), ALIASFILE);
		configfile = line;
	}

	f = fopen(configfile,"r");
	if (!f) {
		err("couldn't open configfile %s: X10CONFIG set?",
			configfile);
		exit(1);
	}

	for (;;) {

		line [0] = '\0';
		fgets(line, 255, f);

		if (line[0] == '\0')
			break;

		line[strlen(line)-1] = '\0';

		comment = strchr(line, '#');
		if (comment) *comment = '\0';

		if (i >= MODULES) {
			err("out of table space for config", NULL);
			break;
		}

		n = sscanf(line, "%s %s %s", alias, housecode, unit);

		if (n <= 0)
			continue;

		// printf("cfg: %d: %s\n", n, line);

		switch(n) {
		case 2:
			if (strcmp(alias, "TTY") == 0) {
				extern char x10_tty[];
				/* special case */
				strcpy(x10_tty, housecode);
				continue;
			}
			if (strcmp(alias, "LATITUDE") == 0) {
				/* another special case */
				strncpy(latitude, housecode, 
						sizeof(latitude));
				continue;
			}
			if (strcmp(alias, "LONGITUDE") == 0) {
				/* another special case */
				strncpy(longitude, housecode, 
						sizeof(longitude));
				continue;
			}
			if (strcmp(alias, "HOUSECODE") == 0) {
				/* another special case */
				x10_housecode = char2hc(housecode[0]);
				continue;
			}

			// printf("alias: %s --> %s\n", alias, housecode);
			strncpy(x10_modules[i].alias, housecode, NAME_LEN);
			strncpy(x10_modules[i].name, alias, NAME_LEN);
			++i;
			break;

		case 3:
			housecode[1] = '\0';
			if (isupper(housecode[0]))
				housecode[0] = tolower(housecode[0]);
			if (housecode[0] < 'a' || housecode[0] > 'p') {
				err("bad housecode in config, alias %s",
					alias);
				continue;
			}
			x10_modules[i].hc = housecode[0];
			strncpy(x10_modules[i].un,unit,UNIT_LEN);
			strncpy(x10_modules[i].name, alias, NAME_LEN);

			++i;
			break;

		case 1:
			err("warning: short field in line in config:\n	%s",line);
			break;
		default:
			err("warning: too many fields in line in config:\n	%s",line);
			break;
		}

	}

	fclose (f);

}



struct x10_mod *
xmod_lookup(name)
char *name;
{
	int i = 0;
	struct x10_mod *xm;
	char *lookname = name;

	xm = x10_modules;
	while (i < MODULES) {
		if (!xm->name[0])
			break;
		// printf("checking %s against %s (%p)\n", lookname, xm->name, xm);
		if (strcmp(lookname, xm->name) == 0) {
			if (!xm->alias[0]) {
				return xm;
			} 
			// printf("found alias: %s --> %s\n", lookname, xm->alias);
			lookname = xm->alias;
		}
		i++;
		xm++;
	}
	if (!i) {
		err("no config at all", NULL);
	}
	return NULL;
}

char *
xmod_name(hl,unit)
char *unit;
{
	int i = 0;
	struct x10_mod *xm;

	xm = x10_modules;
	if (isupper(hl))
		hl = tolower(hl);
	while (i < MODULES) {
		if (!xm->name[0])
			break;
		if (hl == xm->hc && strcmp(unit,xm->un) == 0)
			return xm->name;
		i++;
		xm++;
	}
	return NULL;
}


parse_unit(name,hcp,unitp)
char *name;
int *hcp;
char **unitp;
{
    struct x10_mod *x;
    int hletter;
    static unitnum[5];

    if (isalpha(name[0]) && (isdigit(name[1]) || name[1] == '*')) {
	hletter = name[0];
	if (isupper(hletter))
	    hletter = tolower(hletter);
	*unitp = &name[1];
    } else {
	x = xmod_lookup(name);
	if (!x) {
		err("bad alias %s",name);
		quit();
	}
	hletter = x->hc;
	*unitp = x->un;
    }
    *hcp = char2hc(hletter);

}


err(m,s)
char *m, *s;
{
	fprintf(stderr, "x10: ");
	fprintf(stderr, m, s);
	fprintf(stderr, "\n");
}

dump_config()
{
	int i = 0;
	struct x10_mod *xm = x10_modules;
	extern char x10_tty[];
	while (i < MODULES) {
		if (!xm->name[0])
			break;
		printf("%d: name: %s, housecode is %c, unit %s\n",
			i, xm->name, xm->hc, xm->un);
		i++;
		xm++;
	}
	if (!i)
		err("no config at all", NULL);

	if (x10_tty[0])
		printf("tty is %s\n", x10_tty);
}

